File: OS_TOS/client/lib/attr.lua
FileAttr = {}
FileAttr.__index = FileAttr

local BIT_DIRECTORY = 15
local BIT_DELETED = 14
local BIT_SUID = 11
local BIT_SGID = 10
local BIT_STICKY = 9
local BIT_OWNER_R = 8
local BIT_OWNER_W = 7
local BIT_OWNER_X = 6
local BIT_GROUP_R = 5
local BIT_GROUP_W = 4
local BIT_GROUP_X = 3
local BIT_WORLD_R = 2
local BIT_WORLD_W = 1
local BIT_WORLD_X = 0

local function is_set(value, index)
  return (value & (1 << index)) ~= 0
end

local function set(value, index, flag)
  if flag then
    return value | (1 << index)
  else
    return value & ~(1 << index)
  end
end

function FileAttr:create(value)
  value = value or 0
  local data = {
    value = value
  }
  local metatable = {
    __index = self
  }
  return setmetatable(data, metatable)
end

function FileAttr:is_dir()
  return is_set(self.value, BIT_DIRECTORY)
end

function FileAttr:is_file()
  return not self:is_dir()
end

function FileAttr:deleted()
  return not is_set(self.value, BIT_DELETED)
end

function FileAttr:suid()
  return is_set(self.value, BIT_SUID)
end

function FileAttr:sgid()
  return is_set(self.value, BIT_SGID)
end

function FileAttr:sticky()
  return is_set(self.value, BIT_STICKY)
end

function FileAttr:owner(index)
  return is_set(self.value, BIT_OWNER_R - index)
end

function FileAttr:group(index)
  return is_set(self.value, BIT_GROUP_R - index)
end

function FileAttr:world(index)
  return is_set(self.value, BIT_WORLD_R - index)
end

function FileAttr:can(who, what)
  local offset = 2 + 3 * (2 - who)
  return is_set(self.value, offset - what)
end

function FileAttr:set_dir(flag)
  self.value = set(self.value, BIT_DIRECTORY, flag)
  return self
end

function FileAttr:set_file(flag)
  return self:set_dir(not flag)
end

function FileAttr:set_deleted(flag)
  self.value = set(self.value, BIT_DELETED, not flag)
  return self
end

function FileAttr:set_suid(flag)
  self.value = set(self.value, BIT_SUID, flag)
  return self
end

function FileAttr:set_sgid(flag)
  self.value = set(self.value, BIT_SGID, flag)
  return self
end

function FileAttr:set_sticky(flag)
  self.value = set(self.value, BIT_STICKY, flag)
  return self
end

function FileAttr:set_owner(index, flag)
  self.value = set(self.value, BIT_OWNER_R - index, flag)
  return self
end

function FileAttr:set_group(index, flag)
  self.value = set(self.value, BIT_GROUP_R - index, flag)
  return self
end

function FileAttr:set_world(index, flag)
  self.value = set(self.value, BIT_WORLD_R - index, flag)
  return self
end

function FileAttr:set_can(who, what, flag)
  local offset = 2 + 3 * (2 - who)
  self.value = set(self.value, offset - what, flag)
  return self
end

function FileAttr:value()
  return self.value
end

local export = {
  FileAttr=FileAttr,
}
return export

File: OS_TOS/src/fs_lib/src/attr.cpp
#include "attr.h"
#include "bit_util.h"

using namespace fs;
using namespace fs::internal;

#define BIT_DIRECTORY  15
#define BIT_DELETED    14
#define BIT_SUID       11
#define BIT_SGID       10
#define BIT_STICKY     9
#define BIT_OWNER_R    8
#define BIT_OWNER_W    7
#define BIT_OWNER_X    6
#define BIT_GROUP_R    5
#define BIT_GROUP_W    4
#define BIT_GROUP_X    3
#define BIT_WORLD_R    2
#define BIT_WORLD_W    1
#define BIT_WORLD_X    0

FileAttr::FileAttr(const uint16_t raw) : m_raw(raw) {}

FileAttr::FileAttr() : FileAttr(0)
{
}

bool FileAttr::is_dir() const
{
	return is_set(BIT_DIRECTORY);
}

bool FileAttr::is_file() const
{
	return !is_dir();
}

bool FileAttr::deleted() const
{
	return !is_set(BIT_DELETED);
}

bool FileAttr::suid() const
{
	return is_set(BIT_SUID);
}

bool FileAttr::sgid() const
{
	return is_set(BIT_SGID);
}

bool FileAttr::sticky() const
{
	return is_set(BIT_STICKY);
}

bool FileAttr::owner(uint8_t index) const
{
	return is_set(BIT_OWNER_R - index);
}

bool FileAttr::group(uint8_t index) const
{
	return is_set(BIT_GROUP_R - index);
}

bool FileAttr::world(uint8_t index) const
{
	return is_set(BIT_WORLD_R - index);
}

bool FileAttr::can(uint8_t who, uint8_t what) const
{
	uint8_t offset = 2 + 3 * (2 - who);
	return is_set(offset - what);
}

FileAttr& FileAttr::is_dir(bool value)
{
	set(BIT_DIRECTORY, value);
	return *this;
}

FileAttr& FileAttr::is_file(bool value)
{
	return is_dir(!value);
}

FileAttr& FileAttr::deleted(bool value)
{
	set(BIT_DELETED, !value);
	return *this;
}

FileAttr& FileAttr::suid(bool value)
{
	set(BIT_SUID, value);
	return *this;
}

FileAttr& FileAttr::sgid(bool value)
{
	set(BIT_SGID, value);
	return *this;
}

FileAttr& FileAttr::sticky(bool value)
{
	set(BIT_STICKY, value);
	return *this;
}

FileAttr& FileAttr::owner(uint8_t index, bool value)
{
	set(BIT_OWNER_R - index, value);
	return *this;
}

FileAttr& FileAttr::group(uint8_t index, bool value)
{
	set(BIT_GROUP_R - index, value);
	return *this;
}

FileAttr& FileAttr::world(uint8_t index, bool value)
{
	set(BIT_GROUP_R - index, value);
	return *this;
}

FileAttr& FileAttr::can(uint8_t who, uint8_t what, bool value)
{
	uint8_t offset = 2 + 3 * (2 - who);
	set(offset - what, value);
	return *this;
}

const uint16_t FileAttr::value() const
{
	return m_raw;
}

bool FileAttr::is_set(uint8_t index) const
{
	return binary_is_set(m_raw, index);
}

void FileAttr::set(uint8_t index, bool value)
{
	binary_toggle(m_raw, index, value);	
}
